AVX-21060 Sort lists in aviatrix_firewall_instance_images data source

MingweiZhu666 opened the pull request:
Sort lists in aviatrix_firewall_instance_images data source
fix:
sort the firewall_image_version list form latest to old
can sort version format example:
 firewall_image_version = [
                "10.2.0",
                "10.0.9",
                "10.0.8-h8",
                "10.0.7",
            ]
 firewall_image_version = [
                "9.1.2",
                "9.1.0-h3",
                "9.0.9-h1.xfr",
                "9.0.9.xfr",
            ]
 firewall_image_version = [
                "(7.0.5)",
                "(7.0.3)",
                "(6.4.8)",
                "(6.4.7)",
            ]
  firewall_image_version = [
                "R81.10-335.986",
                "R81-392.983",
                "R80.40-294.983",
            ]
  firewall_image_version = [
                "R81.10_rev1.1",
                "R81.10_rev1.0",
                "R81_rev1.2",
            ]
  firewall_image_version = [
                "PA-VM-10.2.0",
                "PA-VM-10.1.0",
                "PA-VM-10.0.9",
            ]
sort the firewall_size list

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix:
sort the firewall_image_version list form latest to old
sort the firewall_size list from small to large

mlin-aviatrix commented on aviatrix/data_source_aviatrix_firewall_instance_images.go:
```diff
@@ -73,3 +80,140 @@ func dataSourceAviatrixFirewallInstanceImagesRead(d *schema.ResourceData, meta i
+
+func sortImageVersion(listVersion []string) []string {
+	var n = len(listVersion)
+	for i := 0; i <= n-1; i++ {
```
I think will be better to use the builtin sort.Slice function instead of iterating over the entire list. It will definitely be faster than n^2 nested for loops (not that it really matters as there will always be a small number of elements in the given slice), but I think it is more clear.

You can learn more about sort.Slice at https://pkg.go.dev/sort#Slice

MingweiZhu666 replied:
thank you! i will check and fix it.

MingweiZhu666 replied:
got your point, have fixed base on the builtin sort.Slice function and custom the less() compare function, thank you!

mlin-aviatrix commented on aviatrix/data_source_aviatrix_firewall_instance_images.go:
```diff
@@ -73,3 +80,140 @@ func dataSourceAviatrixFirewallInstanceImagesRead(d *schema.ResourceData, meta i
+		reg, _ := regexp.Compile("[^0-9.]+")
+		v1SliceString := reg.ReplaceAllString(imageVersionArray1[index], "")
+		v2SliceString := reg.ReplaceAllString(imageVersionArray2[index], "")
+		int1, _ := strconv.ParseFloat(v1SliceString, 32)
```
Is it a good idea to ignore the error here? Can we assume that this will always succeed?

MingweiZhu666 replied:
i will add the error log here, thank you

mlin-aviatrix commented on aviatrix/data_source_aviatrix_firewall_instance_images.go:
```diff
@@ -73,3 +80,140 @@ func dataSourceAviatrixFirewallInstanceImagesRead(d *schema.ResourceData, meta i
+	for i := 0; i <= n-1; i++ {
+		for j := i; j <= n-1; j++ {
+			var splitFlag string
+			if checkFirstCharacter(listVersion[i]) == "R" {
```
Instead of deciding how to sort by using the first character in the version name, you should probably use the description. That way if we add anther type of versioning that starts with R, but has a different format it won't break .

shuoliugit commented:
There is a nice solution from Hashicorp - https://github.com/hashicorp/go-version
https://stackoverflow.com/questions/18409373/how-to-compare-two-version-number-strings-in-golang

Need discussion to decide whether to use this or not.

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix:
add error log

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix:
use the builtin sort.Slice function to sort version and image size

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix:
combine compareVersion and compareVersion2 method

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix:
use go-version library to sort firewall_image_version

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix:
use go-version library to sort firewall_image_version

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source

mlin-aviatrix commented on aviatrix/data_source_aviatrix_firewall_instance_images.go:
```diff
@@ -61,8 +68,12 @@ func dataSourceAviatrixFirewallInstanceImagesRead(d *schema.ResourceData, meta i
-		fI["firewall_image_version"] = image.Version
-		fI["firewall_size"] = image.Size
+		listVersion := image.Version
+		sort.Sort(sort.Reverse(byVersion(listVersion)))
```
You should be able to use sort.Slice instead of sort.Sort. That way, you only need to define 2 new Less functions and you don't have to define new types.

MingweiZhu666 replied:
fix:
sort.Slice(versionList, func(i, j int) bool {
 			return sortVersion(versionList, i, j)
 		})
 fI["firewall_image_version"] = reverseArray(versionList)
 sizeList := image.Size
 sort.Slice(sizeList, func(i, j int) bool {
 			return sortSize(sizeList, i, j)
 		})

mlin-aviatrix commented on aviatrix/data_source_aviatrix_firewall_instance_images.go:
```diff
@@ -73,3 +84,115 @@ func dataSourceAviatrixFirewallInstanceImagesRead(d *schema.ResourceData, meta i
+//Rxx.xx-xxx.xxx
+//Rxx.xx_revx.x
+func versionFormat1(version1, version2, flag string) bool {
+	//[Rxx.xx, xxx.xxx]
```
Not sure what this comment means. Can you be more explicit about what this is trying to say?

MingweiZhu666 replied:
remove this comment and add new comment to descripe the function

mlin-aviatrix commented on aviatrix/data_source_aviatrix_firewall_instance_images.go:
```diff
@@ -73,3 +84,115 @@ func dataSourceAviatrixFirewallInstanceImagesRead(d *schema.ResourceData, meta i
 	d.SetId(vpcId)
 	return nil
 }
+
```
Everything under here can probably me moved to `aviatrix/utils.go`

MingweiZhu666 replied:
yes, move all sort methods into utils.go file, thank you

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix:
move sort method into utils.go file
use sort.Slice instead of sort.Sort

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -196,3 +200,97 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
 	return false
 }
+
+func reverseArray(arr []string) []string {
```
Small nit: Technically should be reverseSlice

MingweiZhu666 replied:
have removed the reverseArray function

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -196,3 +200,97 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+	return arr
+}
+
+// sort the firewall_image_version list
```
Comments about the function should start with the function name. So in this case it might be:
`sortVersion sorts the firewall_image_version list`

MingweiZhu666 replied:
have fixed, thank you

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -196,3 +200,97 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+	return false
+}
+
+// sort the firewall_size list
```
Same as above.

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -196,3 +200,97 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+	return false
+}
+
+// this function can sort the firewall_image_version format like: R81.10-335.883 && R81.10_rev1.0
```
Same as above.

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -196,3 +200,97 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+	return compareVersion(reg.ReplaceAllString(versionArray1[0], ""), reg.ReplaceAllString(versionArray2[0], ""))
+}
+
+// this function can sort the firewall_image_version format like: PA-VM-10.1.0
```
Same as above.

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -196,3 +200,97 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+	return compareVersion(versionArray1[2], versionArray2[2])
+}
+
+// this function can sort the Semantic Version
```
Same as above.

mlin-aviatrix commented on aviatrix/data_source_aviatrix_firewall_instance_images.go:
```diff
@@ -68,12 +62,16 @@ func dataSourceAviatrixFirewallInstanceImagesRead(d *schema.ResourceData, meta i
+		sort.Slice(versionList, func(i, j int) bool {
+			return sortVersion(versionList, i, j)
+		})
+		fI["firewall_image_version"] = reverseArray(versionList)
```
Is this reverse really needed? I think you can change how the sort works so that we don't need to reverse the list after. Probably swapping true and false in `compareVersion` should work but you will need to test that.

MingweiZhu666 replied:
remove the reverseArray function and set the compare somatic version function from LessThan to GreaterThan

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -196,3 +200,97 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+		} else if strings.Contains(versionList[i], "-") {
+			return versionFormat1(versionList[i], versionList[j], "-")
+		} else {
+			log.Printf("need to add a new method sort this version format")
```
This should also include an example of the version.

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -196,3 +200,97 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+
+// sort the firewall_image_version list
+func sortVersion(versionList []string, i, j int) bool {
+	if checkFirstCharacter(versionList[i]) == "R" {
```
We probably should check the name of the firewall provider instead of jus the first character. That way we won't run into issues if a new format is added.

MingweiZhu666 replied:
have fixed base on the image name to choose the sort function

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -249,18 +242,18 @@ func versionFormat1(version1, version2, flag string) bool {
 }
 
-// this function can sort the Semantic Version
+// compareVersion sorts the Semantic Version
```
Small Nit: Should be `compareVersion compares two Semantic Versions`

MingweiZhu666 replied:
have fixed: compareVersion compares two Semantic Versions

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -238,7 +231,7 @@ func sortSize(sizeList []string, i, j int) bool {
 }
 
-// this function can sort the firewall_image_version format like: R81.10-335.883 && R81.10_rev1.0
+// versionFormat1 sorts the firewall_image_version format like: R81.10-335.883 && R81.10_rev1.0
```
Should be `versionFormat1 compares...`

mlin-aviatrix replied:
Also the function name could be more descriptive. Maybe something like `compareCheckpointVersion`

MingweiZhu666 replied:
have fixed: compareCheckpointVersion compares firewall_image_version format like: R81.10-335.883 && R81.10_rev1.0

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -249,18 +242,18 @@ func versionFormat1(version1, version2, flag string) bool {
 }
 
-// this function can sort the firewall_image_version format like: PA-VM-10.1.0
+// versionFormat2 sorts the firewall_image_version format like: PA-VM-10.1.0
```
Should be `versionFormat2 compares...`

mlin-aviatrix replied:
This one could also be more descriptive. Something like `comparePAVersion`

MingweiZhu666 replied:
thank you, have fixed: comparePAVersion compares firewall_image_version format like: PA-VM-10.1.0

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix:
sort version will base on the image name to choose the sort function
add checkVersionFormat and removePoint function to convert version to correct Semantic Version format

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
solve merge conflict

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -198,6 +201,107 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+}
+
+// removePoint removes all points after the second point in a version
+func removePoint(version string) string {
```
It doesn't make sense to remove the dot because some comparisons will be ordered incorrectly. For example, `1.0.1.10` > `1.0.10.1` because when the dot is removed it will compare `1.0.110 > 1.0.101`.  The best solution might be to ignore everything after the third dot. We should talk about this.

MingweiZhu666 replied:
yes, you are correct, my mistake, i ignore this scenario, before i only tested this example:
                "9.0.9-h1.xfr",
                "9.0.9.xfr",
                "8.1.15",
  
let me think a new way to solve this issue

MingweiZhu666 replied:
add a new function: removeAfterThirdDotValue, which removes everything after the third dot in a version

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -198,6 +201,107 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+	return v1.GreaterThan(v2)
+}
+
+// checkVersionFormat removes special characters, only keep character, number, point and minus in a version
```
The Semantic Versioning documentation refers to "." as a dot and "-" as a hyphen so we should probably do the same. Also "character, number" can be replaced with alphanumerics

MingweiZhu666 replied:
fixed as: checkVersionFormat removes special characters, only keep dot, hyphen, alphanumerics in a version

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
add removeAfterThirdPointValue function to remove everything after the third dot in a version

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -198,6 +201,115 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+func compareCheckpointVersion(version1, version2, flag string) bool {
+	versionArray1 := strings.Split(version1, flag)
+	versionArray2 := strings.Split(version2, flag)
+	reg, _ := regexp.Compile("[^0-9.-]+")
```
We can probably define const regexp for each one so that we don't need to compile a new regexp every time. Looks at regexp.MustCompile.

MingweiZhu666 replied:
fix as:
        const DotHyphenNumberRegexp = "[^0-9.-]+"
	reg := regexp.MustCompile(DotHyphenNumberRegexp)

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -198,6 +201,115 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+
+// checkVersionFormat removes special characters, only keep dot, hyphen, alphanumerics in a version
+func checkVersionFormat(version string) string {
+	reg, _ := regexp.Compile("[^a-zA-Z0-9.-]+")
```
Same as above.

MingweiZhu666 replied:
fix as:
        const DotHyphenAlphanumericsRegexp = "[^a-zA-Z0-9.-]+"
	reg := regexp.MustCompile(DotHyphenAlphanumericsRegexp)

mlin-aviatrix commented on aviatrix/utils.go:
```diff
@@ -198,6 +201,115 @@ func DiffSuppressFuncGatewayVpcId(k, old, new string, d *schema.ResourceData) bo
+	imageSizeArray2 := strings.Split(imageSize2, flag)
+	for index := range imageSizeArray1 {
+		if index >= indexFlag {
+			reg, _ := regexp.Compile("[^0-9]+")
```
Same as above.

MingweiZhu666 replied:
fix as:
        const NumberRegexp = "[^0-9]+"
	reg := regexp.MustCompile(NumberRegexp)
	for index := range imageSizeArray1 {
		if index >= indexFlag {

mlin-aviatrix commented on go.mod:
```diff
@@ -10,6 +10,7 @@ require (
 	github.com/hashicorp/errwrap v1.1.0 // indirect
 	github.com/hashicorp/go-multierror v1.1.0 // indirect
 	github.com/hashicorp/go-uuid v1.0.2 // indirect
+	github.com/hashicorp/go-version v1.4.0 // indirect
```
I think you need to run `go mod tidy`

MingweiZhu666 replied:
done

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix Regexp

MingweiZhu666 committed:
AVX-21060
Sort lists in aviatrix_firewall_instance_images data source
fix Regexp
